### ISSUE TITLE
Installation error: cannot convert ‘double* const’ to ‘hypre_Vector*’ 

### ISSUE THREAD

**Issue by TJFord** [mfem/mfem#1376]
Hello everyone, 

I searched the issues and did not find any answer my question. I am curious if anyone can help me out. 

I install the serial version through `make serial`, seems to be working. However, when i type `make parallel -j`, the error is at the end of the output. 

$ make parallel -j
make -f makefile config MFEM_USE_MPI=YES MFEM_DEBUG=NO \

make[1]: Entering directory '/data3/tan/mfem-4.1'
make -C config all
make[2]: Entering directory '/data3/tan/mfem-4.1/config'
mpicxx -I../../hypre/src/hypre/include get_hypre_version.cpp -o get_hypre_version
./get_hypre_version > get_hypre_version.out
HYPRE version: 21802
rm get_hypre_version get_hypre_version.out
make[2]: Leaving directory '/data3/tan/mfem-4.1/config'

Build destination: <source> [/data3/tan/mfem-4.1]

make[1]: Leaving directory '/data3/tan/mfem-4.1'
make
make[1]: Entering directory '/data3/tan/mfem-4.1'
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/optparser.cpp -o general/optparser.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/mem_manager.cpp -o general/mem_manager.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/hip.cpp -o general/hip.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/communication.cpp -o general/communication.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/tic_toc.cpp -o general/tic_toc.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/array.cpp -o general/array.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/device.cpp -o general/device.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/isockstream.cpp -o general/isockstream.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/error.cpp -o general/error.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/cuda.cpp -o general/cuda.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/stable3d.cpp -o general/stable3d.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/table.cpp -o general/table.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/version.cpp -o general/version.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/binaryio.cpp -o general/binaryio.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/globals.cpp -o general/globals.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/occa.cpp -o general/occa.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/socketstream.cpp -o general/socketstream.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/sets.cpp -o general/sets.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c general/osockstream.cpp -o general/osockstream.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/matrix.cpp -o linalg/matrix.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/complex_operator.cpp -o linalg/complex_operator.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/operator.cpp -o linalg/operator.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/blockmatrix.cpp -o linalg/blockmatrix.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/sparsemat.cpp -o linalg/sparsemat.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/hypre_parcsr.cpp -o linalg/hypre_parcsr.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/blockoperator.cpp -o linalg/blockoperator.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/blockvector.cpp -o linalg/blockvector.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/petsc.cpp -o linalg/petsc.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/densemat.cpp -o linalg/densemat.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/sparsesmoothers.cpp -o linalg/sparsesmoothers.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/ginkgo.cpp -o linalg/ginkgo.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/hypre.cpp -o linalg/hypre.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/sundials.cpp -o linalg/sundials.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/ode.cpp -o linalg/ode.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/hiop.cpp -o linalg/hiop.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/superlu.cpp -o linalg/superlu.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/solvers.cpp -o linalg/solvers.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/strumpack.cpp -o linalg/strumpack.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/vector.cpp -o linalg/vector.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c linalg/handle.cpp -o linalg/handle.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/pmesh.cpp -o mesh/pmesh.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/ncmesh.cpp -o mesh/ncmesh.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/mesh.cpp -o mesh/mesh.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/triangle.cpp -o mesh/triangle.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/vertex.cpp -o mesh/vertex.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/segment.cpp -o mesh/segment.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/wedge.cpp -o mesh/wedge.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/mesh_operators.cpp -o mesh/mesh_operators.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/quadrilateral.cpp -o mesh/quadrilateral.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/mesh_readers.cpp -o mesh/mesh_readers.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/element.cpp -o mesh/element.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/pncmesh.cpp -o mesh/pncmesh.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/pumi.cpp -o mesh/pumi.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/nurbs.cpp -o mesh/nurbs.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/tetrahedron.cpp -o mesh/tetrahedron.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/point.cpp -o mesh/point.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/mesquite.cpp -o mesh/mesquite.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/vtk.cpp -o mesh/vtk.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c mesh/hexahedron.cpp -o mesh/hexahedron.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/plinearform.cpp -o fem/plinearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/gridfunc.cpp -o fem/gridfunc.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/pfespace.cpp -o fem/pfespace.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/nonlinearform_ext.cpp -o fem/nonlinearform_ext.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilinearform.cpp -o fem/bilinearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/quadinterpolator_face.cpp -o fem/quadinterpolator_face.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_divergence.cpp -o fem/bilininteg_divergence.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/geom.cpp -o fem/geom.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/pbilinearform.cpp -o fem/pbilinearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/nonlininteg.cpp -o fem/nonlininteg.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/hybridization.cpp -o fem/hybridization.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/restriction.cpp -o fem/restriction.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_diffusion.cpp -o fem/bilininteg_diffusion.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_gradient.cpp -o fem/bilininteg_gradient.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_vecdiffusion.cpp -o fem/bilininteg_vecdiffusion.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/coefficient.cpp -o fem/coefficient.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/eltrans.cpp -o fem/eltrans.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_hcurl.cpp -o fem/bilininteg_hcurl.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/gslib.cpp -o fem/gslib.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/conduitdatacollection.cpp -o fem/conduitdatacollection.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/intrules.cpp -o fem/intrules.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/datacollection.cpp -o fem/datacollection.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/fespace.cpp -o fem/fespace.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/fe.cpp -o fem/fe.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/tmop_tools.cpp -o fem/tmop_tools.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_vecmass.cpp -o fem/bilininteg_vecmass.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/fe_coll.cpp -o fem/fe_coll.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/nonlinearform.cpp -o fem/nonlinearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_dgtrace.cpp -o fem/bilininteg_dgtrace.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/prestriction.cpp -o fem/prestriction.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/nonlininteg_vectorconvection.cpp -o fem/nonlininteg_vectorconvection.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilinearform_ext.cpp -o fem/bilinearform_ext.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg.cpp -o fem/bilininteg.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/estimators.cpp -o fem/estimators.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/pnonlinearform.cpp -o fem/pnonlinearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/linearform.cpp -o fem/linearform.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/quadinterpolator.cpp -o fem/quadinterpolator.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_convection.cpp -o fem/bilininteg_convection.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/tmop.cpp -o fem/tmop.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/complex_fem.cpp -o fem/complex_fem.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/staticcond.cpp -o fem/staticcond.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/lininteg.cpp -o fem/lininteg.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/bilininteg_mass.cpp -o fem/bilininteg_mass.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/sidredatacollection.cpp -o fem/sidredatacollection.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/pgridfunc.cpp -o fem/pgridfunc.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/libceed/diffusion.cpp -o fem/libceed/diffusion.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/libceed/mass.cpp -o fem/libceed/mass.o
mpicxx   -O3 -std=c++11 -I./../hypre/src/hypre/include  -c fem/libceed/ceed.cpp -o fem/libceed/ceed.o
**linalg/hypre.cpp: In member function ‘virtual void mfem::HypreSmoother::SetOperator(const mfem::Operator&)’:
linalg/hypre.cpp:1968:59: error: cannot convert ‘double**’ to ‘hypre_Vector*’ for argument ‘4’ to ‘HYPRE_Int hypre_ParCSRComputeL1Norms(hypre_ParCSRMatrix*, HYPRE_Int, HYPRE_Int*, hypre_Vector*)’
       hypre_ParCSRComputeL1Norms(*A, type, NULL, &l1_norms);
                                                           ^
linalg/hypre.cpp: In member function ‘virtual void mfem::HypreSmoother::Mult(const mfem::HypreParVector&, mfem::HypreParVector&) const’:
**linalg/hypre.cpp:2116:39: error: cannot convert ‘double* const’ to ‘hypre_Vector*’ for argument ‘5’ to ‘HYPRE_Int hypre_ParCSRRelax(hypre_ParCSRMatrix*, hypre_ParVector*, HYPRE_Int, HYPRE_Int, hypre_Vector*, HYPRE_Real, HYPRE_Real, HYPRE_Real, HYPRE_Real, HYPRE_Int, HYPRE_Real, hypre_ParVector*, hypre_ParVector*, hypre_ParVector*)’
                            x, *V, NULL);
                                       ^
linalg/hypre.cpp:2121:37: error: cannot convert ‘double* const’ to ‘hypre_Vector*’ for argument ‘5’ to ‘HYPRE_Int hypre_ParCSRRelax(hypre_ParCSRMatrix*, hypre_ParVector*, HYPRE_Int, HYPRE_Int, hypre_Vector*, HYPRE_Real, HYPRE_Real, HYPRE_Real, HYPRE_Real, HYPRE_Int, HYPRE_Real, hypre_ParVector*, hypre_ParVector*, hypre_ParVector*)’
                            x, *V, *Z);****
                                     ^
makefile:419: recipe for target 'linalg/hypre.o' failed
make[1]: *** [linalg/hypre.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/data3/tan/mfem-4.1'
makefile:460: recipe for target 'parallel' failed
make: *** [parallel] Error 2

**Comment by v-dobrev** [comment]
It looks like the `master` version of hypre has changes that are backward incompatible.

Using hypre v2.18.2 or earlier should work fine.

Thank you for reporting this issue.

**Comment by tzanio** [comment]
> It looks like the `master` version of hypre has changes that are backward incompatible.

The change was made [here](https://github.com/hypre-space/hypre/commit/60101610dcadc6f6da1f134deca8501e5d82e2c1#diff-25f5be84fdb76820744f6bd16fc3afda), part of https://github.com/hypre-space/hypre/pull/93

cc: @liruipeng

**Comment by liruipeng** [comment]
Hi, Tzanio,

We changed interface of ams/ads/ame; I will look into how it should work again with mfem.

Thank you for letting us know

-Ruipeng

**Comment by liruipeng** [comment]
I am working on this issue. Question: hypre now requires the users to call HYPRE_Init() and HYPRE_Finalize() before and after all the other HYPRE calls. Where to put these in MFEM?

**Comment by v-dobrev** [comment]
Hmm, are these calls required for non-GPU hypre builds?

If they are required just for GPU builds, then we do not need to call them for now -- mfem can only use non-GPU hypre builds.

If they are always required, then we'll need to think about the options. Ideally, we will want to hide them from mfem users, so that existing examples/miniapps in mfem, and external applications do not need to change anything.

**Comment by tzanio** [comment]
Hi @liruipeng,

Thanks for looking at this and for letting us know of `HYPRE_Init()` / `HYPRE_Finalize()`.

As Veselin said, this will be tricky to handle in mfem -- hypre objects are often created internally and it may not be easy to predict if users will need them or not.

We do have similar calls, `MFEMInitializePets` and `MFEMFinalizePetsc` for PETSc, but hypre is so integrated that adding required calls like this will be very disruptive (i.e. most of the examples and user codes will need to change).

Can you think of a workaround, where this is handled internally in hypre or maybe there is no penalty to call the init multiple times, or some other way where we don't have to expose it to the users?

Cheers,
Tzanio

**Comment by mlstowell** [comment]
Perhaps a singleton.  In the constructor of each HYPRE class we could grab the instance of the singleton and instruct it to initialize HYPRE.  The singleton could contain a single flag which indicates wether or not `HYPRE_Init` has been called.  If it has already been initialized this operation could immediately return.  The destructor of the singleton could call `HYPRE_Finalize`.

The only complication I see is that `HYPRE_Init` needs `argc` and `argv` (if I found the right declaration) so we might need to pass these to every HYPRE object.  An alternative might be to pass dummy versions of these if we know exactly what HYPRE needs from them.

**Comment by liruipeng** [comment]
working on a workaround that can avoid `HYPRE_Init` for non-GPU builds

**Comment by tzanio** [comment]
This specific error reported here will be addressed in https://github.com/hypre-space/hypre/pull/102.

@liruipeng, can we please continue the  `HYPRE_Init()` discussion in https://github.com/mfem/mfem/issues/1276?

**Comment by liruipeng** [comment]
> The specific error reported here will be addressed in [hypre-space/hypre#102](https://github.com/hypre-space/hypre/pull/102).
> 
> @liruipeng, can we please continue the `HYPRE_Init()` discussion in #1276?

Part of `HYPRE_Init()` is to create a global struct called `_hypre_handle`, where a bunch of hypre's parameters and global data are stored. For a non-GPU build, now if 'HYPRE_Init()` is not called as the first `HYPRE` call,  `_hypre_handle` will be created at the first time it is accessed (done with the recent PR of hypre). With that being said, the code will not seg fault without calling `HYPRE_Init()`.

**Comment by TJFord** [comment]
Thank you all for the help! @v-dobrev @liruipeng @tzanio @mlstowell 
It worked great in both serial and parallel versions with ex1 and ex1p. 
Here is what I did. I write it down for my own future reference. Hopefully it would be helpful for others as well. 
in hypre folder: 
```
git pull
./configure --disable-fortran
make -j
```
I did not use the following section, as the folder I check out is hypre without version number. 
```
cd ../..
~> ln -s hypre-2.16.0 hypre
```
Then I go to metis:
```
 cd metis-5.1.0
 make config ; make
 mkdir lib
 ln -s ../build/Linux-x86_64/libmetis/libmetis.a lib
```
Initially I did not put `..` in the last line, then I got the error with cannot link -lmetis, similar to #1196, 
Then, I found issue #663 where I change the settings in `MFEM/configure/defaults.mk`, including `MFEM_USE_MPI, MFEM_USE_METIS_5, METIS_DIR=@MFEM_DIR@../metis-5.1.0`

next, in MFEM directory,
```
make config
cd examples
make ex1
make ex1p
mpirun -np 4 ex1p
```

it worked! Thank you!
